pkg/operators/wasm/maps.go

```diff
@@ -0,0 +1,208 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+
+	"github.com/cilium/ebpf"
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+func (i *wasmOperatorInstance) addMapFuncs(env wazero.HostModuleBuilder) {
+	exportFunction(env, "getMap", i.getMap,
+		[]wapi.ValueType{wapi.ValueTypeI64}, // MapName
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Map
+	)
+
+	exportFunction(env, "mapLookup", i.mapLookup,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+			wapi.ValueTypeI64, // Value pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "mapUpdate", i.mapUpdate,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+			wapi.ValueTypeI64, // Value pointer
+			wapi.ValueTypeI64, // Flag
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "mapDelete", i.mapDelete,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+}
+
+// getMap gets an existing map.
+// Params:
+// - stack[0] is the name of the map (string encoded)
+// Return value:
+// - Map handle on success, 0 on error
+func (i *wasmOperatorInstance) getMap(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapNamePtr := stack[0]
+
+	mapName, err := stringFromStack(m, mapNamePtr)
+	if err != nil {
+		i.logger.Warnf("getMap: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok := i.gadgetCtx.GetVar(operators.MapPrefix + mapName)
+	if !ok {
+		i.logger.Warnf("get map: no map for name %q", mapName)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok = ebpfMap.(*ebpf.Map)
+	if !ok {
+		i.logger.Warnf("get map: map %q is not an ebpf map", mapName)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = wapi.EncodeU32(i.addHandle(ebpfMap))
+}
+
+// mapLookup searches the map for a value corresponding to the given key.
+// Params:
+// - stack[0]: Map handle
+// - stack[1]: Key pointer
+// - stack[2]: Value pointer
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) mapLookup(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapHandle := wapi.DecodeU32(stack[0])
+	keyPtr := stack[1]
+	valuePtr := stack[2]
+
+	ebpfMap, ok := getHandle[*ebpf.Map](i, mapHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	key, err := bufFromStack(m, keyPtr)
+	if err != nil {
+		i.logger.Warnf("mapLookup: getting a buf for key pointer: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	value, err := ebpfMap.LookupBytes(key)
```

Comment:
Similarly to `mapUpdate`, we should check the type of the map. For map types that would return a file descriptor, this would cause the kernel to add a file descriptor in the ig process. For now we can support arrays and hash tables. Other map types can be added later if needed.

Reply:
My issue was that the wasm operator would not know that a fd is added to the ig process, so it would never close it. So the process would leak file descriptors, even after the gadget is stopped.

I tried `eventsMap.Lookup` on a perf event array map (simulating an old kernel to avoid using the bpf ring buffer). It fails with ENOTSUPP:
```
[pid 76443] bpf(BPF_MAP_LOOKUP_ELEM, {map_fd=33, key=0xc003ad41e0, value=0xc005275f90, flags=BPF_ANY}, 32) = -1 ENOTSUPP (Unknown error 524)
```
See https://github.com/torvalds/linux/blob/v6.8/kernel/bpf/arraymap.c#L818

So it is not too bad for this map kind. But for maps of maps, this is still a problem.

I also tried to call Update:
```
		eventsMap, err := api.GetMap("events")
		if err != nil {
			api.Warnf("failed to get events map: %s", err)
			return
		}
		var zero [4]byte
		fd := uint32(44)
		err = eventsMap.Update(&zero, &fd, 0)
		if err != nil {
			api.Warnf("failed to lookup events map: %s", err)
			return
		}
```

Strace shows it fails with EPERM:
```
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=13, key=0xc00394f620, value=0xc00394f670, flags=BPF_ANY}, 32) = -1 EPERM (Operation not permitted)
```

This is because the map was frozen in #3151.

But for maps of maps, that should work.

I don't think it is appropriate for the wasm program to poke around ig file descriptor by knowing the fd integer, rather than using a handle.